# Homework 7

Use this notebook to work on your answers and check solutions. You can then submit your functions using "HW7_submission.ipynb" or directly write your functions in a file named "hw7_answers.py".

You will use the cereal dataset attached with the assignment.

In [12]:
import pandas as pd
import numpy as np

## Question 1

Write a function called "get_corrs" which takes one argument:

* df, which is a pandas data frame

and returns:

* m, a correlation matrix for the numerical variables in df.

Hint:
* google _df corr_

In [14]:
cereal = pd.read_csv("~/Desktop/DataProg/cereal.csv")

In [15]:
cereal

name mfr type  calories  protein  fat  \
0                                100% Bran   N    C        70        4    1   
1                        100% Natural Bran   Q    C       120        3    5   
2                                 All-Bran   K    C        70        4    1   
3                All-Bran with Extra Fiber   K    C        50        4    0   
4                           Almond Delight   R    C       110        2    2   
5                  Apple Cinnamon Cheerios   G    C       110        2    2   
6                              Apple Jacks   K    C       110        2    0   
7                                  Basic 4   G    C       130        3    2   
8                                Bran Chex   R    C        90        2    1   
9                              Bran Flakes   P    C        90        3    0   
10                            Cap'n'Crunch   Q    C       120        1    2   
11                                Cheerios   G    C       110        6    2   
12                   Cinnamon Toast Crunch   G    C       120        1    3   
13                                Clusters   G    C       110        3    2   
14                             Cocoa Puffs   G    C       110        1    1   
15                               Corn Chex   R    C       110        2    0   
16                             Corn Flakes   K    C       100        2    0   
17                               Corn Pops   K    C       110        1    0   
18                           Count Chocula   G    C       110        1    1   
19                      Cracklin' Oat Bran   K    C       110        3    3   
20                  Cream of Wheat (Quick)   N    H       100        3    0   
21                                 Crispix   K    C       110        2    0   
22                  Crispy Wheat & Raisins   G    C       100        2    1   
23                             Double Chex   R    C       100        2    0   
24                             Froot Loops   K    C       110        2    1   
25                          Frosted Flakes   K    C       110        1    0   
26                     Frosted Mini-Wheats   K    C       100        3    0   
27  Fruit & Fibre Dates, Walnuts, and Oats   P    C       120        3    2   
28                           Fruitful Bran   K    C       120        3    0   
29                          Fruity Pebbles   P    C       110        1    1   
..                                     ...  ..  ...       ...      ...  ...   
47                    Multi-Grain Cheerios   G    C       100        2    1   
48                        Nut&Honey Crunch   K    C       120        2    1   
49               Nutri-Grain Almond-Raisin   K    C       140        3    2   
50                       Nutri-grain Wheat   K    C        90        3    0   
51                    Oatmeal Raisin Crisp   G    C       130        3    2   
52                   Post Nat. Raisin Bran   P    C       120        3    1   
53                              Product 19   K    C       100        3    0   
54                             Puffed Rice   Q    C        50        1    0   
55                            Puffed Wheat   Q    C        50        2    0   
56                      Quaker Oat Squares   Q    C       100        4    1   
57                          Quaker Oatmeal   Q    H       100        5    2   
58                             Raisin Bran   K    C       120        3    1   
59                         Raisin Nut Bran   G    C       100        3    2   
60                          Raisin Squares   K    C        90        2    0   
61                               Rice Chex   R    C       110        1    0   
62                           Rice Krispies   K    C       110        2    0   
63                          Shredded Wheat   N    C        80        2    0   
64                  Shredded Wheat 'n'Bran   N    C        90        3    0   
65               Shredded Wheat spoon size   N    C        90        3    0   
66       

In [16]:
#### play with code here #####
# We can use the default method from Pandas dataframe.corr()

def get_corrs(dataframe):
    corr_max = dataframe.corr(method = "pearson")
    return corr_max

In [17]:
get_corrs(cereal)

calories   protein       fat    sodium     fiber     carbo  \
calories  1.000000  0.019066  0.498610  0.300649 -0.293413  0.250681   
protein   0.019066  1.000000  0.208431 -0.054674  0.500330 -0.130864   
fat       0.498610  0.208431  1.000000 -0.005407  0.016719 -0.318043   
sodium    0.300649 -0.054674 -0.005407  1.000000 -0.070675  0.355983   
fiber    -0.293413  0.500330  0.016719 -0.070675  1.000000 -0.356083   
carbo     0.250681 -0.130864 -0.318043  0.355983 -0.356083  1.000000   
sugars    0.562340 -0.329142  0.270819  0.101451 -0.141205 -0.331665   
potass   -0.066609  0.549407  0.193279 -0.032603  0.903374 -0.349685   
vitamins  0.265356  0.007335 -0.031156  0.361477 -0.032243  0.258148   
shelf     0.097234  0.133865  0.263691 -0.069719  0.297539 -0.101790   
weight    0.696091  0.216158  0.214625  0.308576  0.247226  0.135136   
cups      0.087200 -0.244469 -0.175892  0.119665 -0.513061  0.363932   
rating   -0.689376  0.470618 -0.409284 -0.401295  0.584160  0.052055   
ratingID -0.661740  0.092726 -0.387242 -0.533138  0.410324 -0.071491   

            sugars    potass  vitamins     shelf    weight      cups  \
calories  0.562340 -0.066609  0.265356  0.097234  0.696091  0.087200   
protein  -0.329142  0.549407  0.007335  0.133865  0.216158 -0.244469   
fat       0.270819  0.193279 -0.031156  0.263691  0.214625 -0.175892   
sodium    0.101451 -0.032603  0.361477 -0.069719  0.308576  0.119665   
fiber    -0.141205  0.903374 -0.032243  0.297539  0.247226 -0.513061   
carbo    -0.331665 -0.349685  0.258148 -0.101790  0.135136  0.363932   
sugars    1.000000  0.021696  0.125137  0.100438  0.450648 -0.032358   
potass    0.021696  1.000000  0.020699  0.360663  0.416303 -0.495195   
vitamins  0.125137  0.020699  1.000000  0.299262  0.320324  0.128405   
shelf     0.100438  0.360663  0.299262  1.000000  0.190762 -0.335269   
weight    0.450648  0.416303  0.320324  0.190762  1.000000 -0.199583   
cups     -0.032358 -0.495195  0.128405 -0.335269 -0.199583  1.000000   
rating   -0.759675  0.380165 -0.240544  0.025159 -0.298124 -0.203160   
ratingID -0.463098  0.239857 -0.300203  0.054436 -0.360595 -0.130983   

            rating  ratingID  
calories -0.689376 -0.661740  
protein   0.470618  0.092726  
fat      -0.409284 -0.387242  
sodium   -0.401295 -0.533138  
fiber     0.584160  0.410324  
carbo     0.052055 -0.071491  
sugars   -0.759675 -0.463098  
potass    0.380165  0.239857  
vitamins -0.240544 -0.300203  
shelf     0.025159  0.054436  
weight   -0.298124 -0.360595  
cups     -0.203160 -0.130983  
rating    1.000000  0.753629  
ratingID  0.753629  1.000000

In [18]:
get_corrs(cereal[['name','calories','carbo','sugars']])

calories     carbo    sugars
calories  1.000000  0.250681  0.562340
carbo     0.250681  1.000000 -0.331665
sugars    0.562340 -0.331665  1.000000

### Sample output:

```
In [1]: get_corrs(cer[['name','calories','carbo','sugars']])
Out[1]:  	calories 	carbo 	    sugars
   calories	1.000000    0.250681 	0.562340
      carbo    0.250681    1.000000    -0.331665
     sugars	0.562340   -0.331665	 1.000000
```

## Question 2

Write a function called "get_corr_pairs" which takes one argument:

* df, which is a pandas data frame

and returns:

* corr_pairs, a dictionary where keys are names of columns of df corresponding to numerical features, and values are arrays of names of columns whose correlation coefficient with the key has magnitude 0.3 or greater. 

You can use your function from question 1 to get the correlation values.

In [19]:
#### play with code here #####
from pprint import pprint
def get_corr_pairs(dataframe):
    corr_max = get_corrs(dataframe)
    col_list = list(corr_max.columns)
    pair_dict = {}
    pair_dict = {col: None for col in col_list}
    for col in col_list:
        pair_list = []
        for i in range(0, len(col_list)):
            if abs(corr_max.loc[col][i]) > 0.3 and abs(corr_max.loc[col][i]) != 1:
                pair_list.append(col_list[i])
        pair_dict.update({col:pair_list})
    return pprint(pair_dict)


In [20]:
get_corr_pairs(cereal)

{'calories': ['fat', 'sodium', 'sugars', 'weight', 'rating', 'ratingID'],
 'carbo': ['fat', 'sodium', 'fiber', 'sugars', 'potass', 'cups'],
 'cups': ['fiber', 'carbo', 'potass', 'shelf'],
 'fat': ['calories', 'carbo', 'rating', 'ratingID'],
 'fiber': ['protein', 'carbo', 'potass', 'cups', 'rating', 'ratingID'],
 'potass': ['protein', 'fiber', 'carbo', 'shelf', 'weight', 'cups', 'rating'],
 'protein': ['fiber', 'sugars', 'potass', 'rating'],
 'rating': ['calories',
            'protein',
            'fat',
            'sodium',
            'fiber',
            'sugars',
            'potass',
            'ratingID'],
 'ratingID': ['calories',
              'fat',
              'sodium',
              'fiber',
              'sugars',
              'vitamins',
              'weight',
              'rating'],
 'shelf': ['potass', 'cups'],
 'sodium': ['calories', 'carbo', 'vitamins', 'weight', 'rating', 'ratingID'],
 'sugars': ['calories', 'protein', 'carbo', 'weight', 'rating', 'ratingID'],

In [21]:
get_corr_pairs(cereal[['name','fat','sugars','rating']])

{'fat': ['rating'], 'rating': ['fat', 'sugars'], 'sugars': ['rating']}


### Sample output:

```
In [1]: get_corr_pairs(cer[['name','fat','sugars','rating']])
Out[1]: {'fat': ['rating'], 'rating': ['fat', 'sugars'], 'sugars': ['rating']}
```

Short explanation: the correlation between 'fat' and 'rating' is -0.409, 'sugars' and 'rating' is -0.760; the remaining correlations have magnitude < 0.3. 

## Question 3

Write a function called "sample_cereal" which takes two arguments:
    
* df, which is a pandas data frame
* kind, which is a string that can take value 'up' or 'down'

and returns:

* a pandas data frame with balanced target class 'ratingID', using up sampling if kind='up' and downsampling if kind='down'.

In [22]:
#### play with code here #####

def sample_cereal(dataframe, kind):
    if kind == "up":
        count = dataframe.apply(lambda x: True if x['ratingID'] == 1 else False, axis=1)
        count1 = len(count[count == True].index)
        count0 = len(count[count == False].index)
        if count1 > count0:
            #major group = rating 1
            #minor group = rating 0
            #we need to up minor group
            #we will do random duplicating from minor group
            diff = count1 - count0
            #diff = how many duplicates we need
            df0 = dataframe.loc[dataframe["ratingID"] == 0]
            random0 = df0.sample(n = diff, random_state = 1)
            #append random0 to dataframe
            return dataframe.append(random0)
        elif count1 < count0:
            #major = rating0
            # up the minor rating 1 group
            diff = count0 - count1
            #find minor group
            df1 = dataframe.loc[dataframe["ratingID"] == 1]
            #random sample minor group
            random1 = df1.sample(n = diff, random_state = 1)
            #append random1 to dataframe
            return dataframe.append(random1)
        elif count1 == count0:
            print ("The dataset is in balance")
    elif kind == "down":
        # down sampling
        count = dataframe.apply(lambda x: True if x['ratingID'] == 1 else False, axis=1)
        count1 = len(count[count == True].index)
        count0 = len(count[count == False].index)
        if count1 > count0:
            #major = rating1
            #drop random rows in rating1
            diff = count1 - count0
            df1 = dataframe.loc[dataframe["ratingID"] == 1]
            random1 = df1.sample(n = diff, random_state = 1)
            random1_list = list(random1.index.values)
            return dataframe.drop(random1_list)
            
        elif count1 < count0:
            #major = rating0
            #drop random rows in rating0
            diff = count0 - count1
            df0 = dataframe.loc[dataframe["ratingID"] == 0]
            random0 = df0.sample(n = diff, random_state = 1)
            random0_list = list(random0.index.values)
            return dataframe.drop(random0_list)
        elif count1 == count0:
            print ("The dataset is in balance.")

In [23]:
sample_cereal(cereal.ix[3:5,['name','mfr','type','calories','protein','ratingID']], 'up')

/Users/xiaofeixue/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:1: DeprecationWarning: 
.ix is deprecated. Please use
.loc for label based indexing or
.iloc for positional indexing

See the documentation here:
http://pandas.pydata.org/pandas-docs/stable/indexing.html#ix-indexer-is-deprecated
  """Entry point for launching an IPython kernel.


name mfr type  calories  protein  ratingID
3  All-Bran with Extra Fiber   K    C        50        4         1
4             Almond Delight   R    C       110        2         0
5    Apple Cinnamon Cheerios   G    C       110        2         0
3  All-Bran with Extra Fiber   K    C        50        4         1

### Sample output:

```
In [1]: sample_cereal(cer.ix[3:5,['name','mfr','type','calories','protein','ratingID']], 'up')
Out[1]:  	name 	            mfr 	type 	calories 	protein 	ratingID
3 	All-Bran with Extra Fiber 	K 	C 	     50 	      4 	     1
3 	All-Bran with Extra Fiber 	K 	C 	     50 	      4 	     1
4 	Almond Delight  	          R 	C 	     110 	     2 	     0
5 	Apple Cinnamon Cheerios 	  G 	C 	     110 	     2 	     0
```

Short explanation: The input has only one positive sample and two negative samples; random sampling from a distribution of 1 can only return one possible result, so our up-sampling of the smaller class merely replicates the row for "All-Bran with Extra Fiber". 

## Question 4

Write a function called "find_H" which takes two arguments:

* df, which is a pandas data frame
* cname, which is the name of the target column (that should correspond to a categorical variable)

and returns:

* H, the entropy in the column cname (use logarithm base 2)

In [119]:
#### play with code here #####
import math
def find_H(dataframe, cname):
    count_bool = dataframe.apply(lambda x: True if x[cname] == 1 else False, axis=1)
    count0 = len(count_bool[count_bool == False].index)
    count1 = len(count_bool[count_bool == True].index)
    x, y = dataframe.shape
    if count0 == 0 or count1 == 0:
        return 0
    else:
        p0 = count0/x
        p1 = count1/x
        ent = -p0*math.log2(p0)-p1*math.log2(p1)
    return ent

In [120]:
cereal = pd.read_csv("~/Desktop/DataProg/cereal.csv")
find_H(cereal.iloc[:20], 'ratingID')

0.6098403047164004

### Sample output:

```
In [1]: find_H(cer.iloc[:20], 'ratingID')
Out[1]: 0.60984030471640038
```

## Question 5

We talked about using gini score for split in class, but we will focus on using entropy in this practice. 
What want to know how much information can be gained by splitting on one specific feature. 

Formally, given resp. variable $Y$, dataset $D$ and feature $X$, information gain of $Y$ from $X$ is defined as 

$ IG(Y, X) = H(Y) - H(Y|X) $

where $H(Y|X)$ is defined as

$H(Y|X) = \sum_{l \in levels(X)} P(l) \cdot H(l). $

See [here](https://medium.com/@rishabhjain_22692/decision-trees-it-begins-here-93ff54ef134) for an illustration.

Write a function called "info_gain" which takes four arguments:

* df, which is a pandas data frame
* cname, which is the name of the target column (that should correspond to a categorical variable)
* csplit, which is the name of a numeric column in df (aka, feature)
* threshold, which is a numeric value (aka, split cutoff)

and returns:

* info_gain, the information gain you get in column cname by splitting the dataset on the threshold value in column csplit. 

In [141]:
#### play with code here #####
def info_gain(dataframe, cname, csplit, threshold):
    threshold = float(threshold)
    x, y = dataframe.shape
    
    master_e = find_H(dataframe, cname)
    
    threshold_up = dataframe[csplit] >= threshold
    up_e = find_H(dataframe[threshold_up], cname)*len(dataframe[threshold_up])/x
    
    threshold_down = dataframe[csplit] < threshold
    down_e = find_H(dataframe[threshold_down], cname)*len(dataframe[threshold_down])/x
    
    gain_info = master_e - up_e - down_e
    return gain_info

In [142]:
cereal = pd.read_csv("~/Desktop/DataProg/cereal.csv")

In [143]:
info_gain(cer.iloc[:20], 'ratingID', 'sugars', '7.0')

0.22806670354641434

### Sample output:


```
In [1]: info_gain(cer.iloc[:20], 'ratingID', 'sugars', '7.0')
Out[1]: 0.2280667035464144
```

Note: in case you wonder, for a probability of 0, use the fact that $lim_{p->0+} p log(p) = 0$. 